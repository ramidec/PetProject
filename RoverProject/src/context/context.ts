import {createContext} from 'react';
import {PetData} from '../utils/const';

export type Context = {
  petData: PetData[];
  setPetData: (petData: PetData[]) => void;
  hasChangedState: boolean;
  setHasChangedState: (changed: boolean) => void;
  refetch: () => void;
};

export const AppContext = createContext<Context | undefined>(undefined);
