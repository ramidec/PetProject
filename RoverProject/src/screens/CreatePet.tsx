import {NativeStackScreenProps} from '@react-navigation/native-stack';
import React, {useCallback, useState} from 'react';
import {
  SafeAreaView,
  View,
  StyleSheet,
  Image,
  ImageSourcePropType,
  Keyboard,
  TouchableOpacity,
} from 'react-native';
import {Button} from '../components/Button';
import {Input} from '../components/Input';
import {colors} from '../utils/colors';
import {useSetPetInAsyncStorage} from '../hooks/storage';
import {useGalleryFiles} from '../hooks/gallery';
import {titles} from '../utils/';

export const CreatePet = ({navigation}: NativeStackScreenProps<{}>) => {
  const [petName, setPetName] = useState<string>('');
  const [petDescription, setPetDescription] = useState<string>('');
  const [saveData] = useSetPetInAsyncStorage();
  const onNameTextChange = useCallback(t => setPetName(t), []);
  const onDescriptionTextChange = useCallback(t => setPetDescription(t), []);

  const [openGallery, imageUri] = useGalleryFiles({
    mediaType: 'photo',
    maxHeight: 390,
    maxWidth: 844,
  });

  const inputData = [
    {
      title: titles.nameInputTitle,
      onTextChange: onNameTextChange,
      placeholder: titles.namePlaceholderText,
    },
    {
      title: titles.descriptionInputTitle,
      onTextChange: onDescriptionTextChange,
      style: {marginTop: 20},
      inputStyle: {height: 125, textAlignVertical: 'top'},
      multiline: true,
      placeholder: titles.descriptionPlaceholderText,
    },
  ];

  const dismissKeyboard = useCallback(() => Keyboard.dismiss(), []);

  const onSubmit = useCallback(() => {
    saveData({
      name: petName,
      description: petDescription,
      petImage: imageUri,
    }).then(() => navigation.goBack());
  }, [navigation, petDescription, imageUri, petName, saveData]);

  return (
    <SafeAreaView style={styles.background}>
      <TouchableOpacity
        activeOpacity={1}
        style={StyleSheet.absoluteFill}
        onPress={dismissKeyboard}>
        <View style={styles.topContainer}>
          {imageUri && (
            <Image
              resizeMode="contain"
              source={imageUri as ImageSourcePropType}
              style={styles.image}
            />
          )}
          <Button
            text={imageUri ? '+' : '+ Upload Photo'}
            color={imageUri ? colors.white : colors.darkGrey}
            style={imageUri ? styles.overlayButton : styles.button}
            onPress={openGallery as () => void}
          />
        </View>
        <View style={styles.inputs}>
          {inputData.map(item => (
            <Input
              onSubmit={dismissKeyboard}
              title={item.title}
              onTextChange={item.onTextChange}
              style={item?.style}
              inputStyle={item?.inputStyle}
              key={item.title}
              multiline={item?.multiline}
              placeholder={item?.placeholder}
            />
          ))}
        </View>
        <Button
          disabled={!petName || !petDescription}
          textStyle={styles.submitButtonText}
          text={titles.submitButtonText}
          color={colors.green}
          style={styles.submitButtom}
          onPress={onSubmit}
        />
      </TouchableOpacity>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  background: {
    flex: 1,
    backgroundColor: colors.white,
  },
  image: {flex: 1, width: '100%'},
  topContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.backgroundGrey,
  },
  submitButtonText: {fontWeight: '700', fontSize: 18},
  submitButtom: {
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
    borderRadius: 20,
    borderWidth: 2,
    width: '90%',
    minHeight: '7.5%',
    backgroundColor: colors.white,
    marginBottom: 20,
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 20,
    borderColor: colors.grey,
    borderWidth: 2,
    width: '40%',
    height: '30%',
    backgroundColor: colors.white,
  },
  overlayButton: {
    position: 'absolute',
    right: 20,
    bottom: 20,
    backgroundColor: colors.green,
    borderRadius: 30,
    height: 30,
    width: 30,
    alignItems: 'center',
    justifyContent: 'center',
  },
  inputs: {
    flex: 3,
    backgroundColor: colors.white,
    paddingTop: 28,
    alignItems: 'center',
  },
});
