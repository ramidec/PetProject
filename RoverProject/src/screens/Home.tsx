import React, {useCallback} from 'react';
import {SafeAreaView, StyleSheet, View, FlatList} from 'react-native';
import {CommonActions} from '@react-navigation/native';

import {EmptyState} from '../components/EmptyState';
import {PetItem} from '../components/PetItem';
import {colors} from '../utils/colors';
import {navigationKeys} from '../utils/const';
import {useDeleteItemFromStorage} from '../hooks/storage';
import {useAppContext} from '../hooks/contexts';
import {NativeStackScreenProps} from '@react-navigation/native-stack';

export const Home = ({navigation}: NativeStackScreenProps<{}>) => {
  const [clearData] = useDeleteItemFromStorage();
  const {petData} = useAppContext();

  const deleteAndRefetch = useCallback(
    id => {
      clearData(id);
    },
    [clearData],
  );

  const renderPetItem = useCallback(
    ({item}) => (
      <PetItem
        petDescription={item?.description}
        petName={item?.name}
        imageSource={item?.petImage}
        id={item?.id}
        deleteAndRefetch={deleteAndRefetch}
      />
    ),
    [deleteAndRefetch],
  );

  const keyExtractor = useCallback((_, index) => index.toString(), []);

  const navigateToCreate = useCallback(() => {
    navigation.dispatch(
      CommonActions.navigate({
        name: navigationKeys.CREATE_PET,
      }),
    );
  }, [navigation]);

  const emptyText =
    'Looks like you havenâ€™t added any pets yet. You can add your first pet profile from the button below.';
  return (
    <SafeAreaView style={styles.background}>
      {petData?.length ? (
        <FlatList
          data={petData}
          extraData={petData}
          renderItem={renderPetItem}
          keyExtractor={keyExtractor}
        />
      ) : (
        <View style={styles.addContainer}>
          <EmptyState text={emptyText} onPress={navigateToCreate} />
        </View>
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  background: {
    flex: 1,
  },
  addContainer: {
    backgroundColor: colors.white,
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
  contentContainer: {
    flex: 1,
    alignItems: 'center',
  },
});
