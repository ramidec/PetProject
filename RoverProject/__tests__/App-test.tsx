/**
 * @format
 */

import 'react-native';
import React from 'react';
import {create} from 'react-test-renderer';
import {Home} from '../src/screens/Home';
import {AppContext, Context} from '../src/context/context';
import {CreatePet} from '../src/screens/CreatePet';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RouteProp} from '@react-navigation/native';
import {
  createID,
  filterData,
  getDataFromStorage,
  getIds,
  hasDuplicates,
  parsePetData,
  setInStorage,
} from '../src/hooks/storage';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {
  ID_TO_DELETE,
  mockedDog,
  mockedInputData,
  mockedPetArray,
  PetData,
} from '../src/utils/const';

beforeEach(async () => {
  await AsyncStorage.clear();
});

const contextMock: Context = {
  petData: [],
  setPetData: () => {},
  hasChangedState: false,
  setHasChangedState: () => {},
  refetch: () => {},
};
let navigationMock;
let routeMock;

const homeTree = create(
  <AppContext.Provider value={contextMock}>
    <Home
      navigation={navigationMock as unknown as NativeStackNavigationProp<{}>}
      route={routeMock as unknown as RouteProp<{}>}
    />
  </AppContext.Provider>,
);
const createPetTree = create(
  <AppContext.Provider value={contextMock}>
    <CreatePet
      navigation={navigationMock as unknown as NativeStackNavigationProp<{}>}
      route={routeMock as unknown as RouteProp<{}>}
    />
  </AppContext.Provider>,
);

describe('screenshots', () => {
  test('home screen snapshot', () => {
    expect(homeTree).toMatchSnapshot();
  });

  test('createPet screen snapshot', () => {
    expect(createPetTree).toMatchSnapshot();
  });
});

describe('test async storage methods', () => {
  test('if no results exist, return an empty array', async () => {
    const result = await getDataFromStorage();
    expect(result).toEqual([]);
  });

  test('if you create a new pet, a unique id is generated for it', async () => {
    let newArray = [...(mockedPetArray as PetData[])];
    const newPet = parsePetData({
      data: mockedInputData,
      id: createID(mockedPetArray),
    });
    newArray.push(newPet as PetData);
    expect(newPet.id).not.toMatch(mockedPetArray[0].id);
    expect(newPet.id).not.toMatch(mockedPetArray[1].id);
    expect(hasDuplicates(getIds(newArray))).toBeFalsy();
  });

  test('if a result was saved, return array with it', async () => {
    setInStorage([mockedDog] as unknown as PetData[]);
    const result = await getDataFromStorage();
    expect(result).toHaveLength(1);
    expect(result[0]).toEqual(mockedDog);
  });

  test('if one result is deleted, return updated array', async () => {
    setInStorage(mockedPetArray as unknown as PetData[]);

    const storage = await getDataFromStorage();
    const filteredData = filterData(storage, ID_TO_DELETE);

    setInStorage(filteredData as unknown as PetData[]);

    const filteredStorage = await getDataFromStorage();

    expect(filteredStorage).toHaveLength(1);
    expect(filteredStorage[0]).toEqual(mockedPetArray[1]);
  });
});
