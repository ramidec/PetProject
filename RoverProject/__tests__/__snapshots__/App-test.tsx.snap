// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screenshots createPet screen snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(196, 196, 196, 0.5)",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#CFD7DE",
            "borderRadius": 20,
            "borderWidth": 2,
            "height": "30%",
            "justifyContent": "center",
            "opacity": 1,
            "width": "40%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#66727C",
                "padding": 5,
              },
              undefined,
            ]
          }
        >
          + Upload Photo
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flex": 3,
          "paddingTop": 28,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "width": "90%",
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Object {
              "alignSelf": "flex-start",
              "fontSize": 17,
              "fontWeight": "700",
            }
          }
        >
          Pet Name *
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          blurOnSubmit={true}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="🐶 Your pet name"
          rejectResponderTermination={true}
          returnKeyType="done"
          style={
            Array [
              Object {
                "borderColor": "#C4C4C4",
                "borderRadius": 6,
                "borderWidth": 2,
                "height": 50,
                "marginTop": 10,
                "padding": 10,
                "width": "100%",
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "width": "90%",
            },
            Object {
              "marginTop": 20,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "alignSelf": "flex-start",
              "fontSize": 17,
              "fontWeight": "700",
            }
          }
        >
          Pet Description *
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          blurOnSubmit={true}
          multiline={true}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="🐶 Your pet description"
          rejectResponderTermination={true}
          returnKeyType="done"
          style={
            Array [
              Object {
                "borderColor": "#C4C4C4",
                "borderRadius": 6,
                "borderWidth": 2,
                "height": 50,
                "marginTop": 10,
                "padding": 10,
                "width": "100%",
              },
              Object {
                "height": 125,
                "textAlignVertical": "top",
              },
            ]
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#CFD7DE",
          "borderRadius": 20,
          "borderWidth": 2,
          "justifyContent": "center",
          "marginBottom": 20,
          "minHeight": "7.5%",
          "opacity": 1,
          "width": "90%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#CFD7DE",
              "padding": 5,
            },
            Object {
              "fontSize": 18,
              "fontWeight": "700",
            },
          ]
        }
      >
        + Add Pet
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`screenshots home screen snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingHorizontal": 55,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "500",
            "textAlign": "center",
          }
        }
      >
        Looks like you haven’t added any pets yet. You can add your first pet profile from the button below.
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#CFD7DE",
            "borderRadius": 20,
            "borderWidth": 2,
            "height": 44,
            "justifyContent": "center",
            "marginTop": 30,
            "opacity": 1,
            "width": 181,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#66727C",
                "padding": 5,
              },
              undefined,
            ]
          }
        >
          + Add Your First Pet
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
